mentat/session.py

```diff
@@ -0,0 +1,215 @@
+import asyncio
+import logging
+import shlex
+import traceback
+from pathlib import Path
+from typing import List, Optional, Union, cast
+from uuid import uuid4
+
+from .code_context import CodeContext
+from .code_edit_feedback import get_user_feedback_on_edits
+from .code_file_manager import CodeFileManager
+from .commands import Command
+from .config_manager import ConfigManager
+from .conversation import Conversation
+from .errors import MentatError
+from .git_handler import get_shared_git_root_for_paths
+from .llm_api import CostTracker, setup_api_key
+from .parsers.block_parser import BlockParser
+from .parsers.parser import Parser
+from .parsers.replacement_parser import ReplacementParser
+from .parsers.split_diff_parser import SplitDiffParser
+from .session_input import collect_user_input
+from .session_stream import SessionStream, set_session_stream
+
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+parser_map: dict[str, Parser] = {
+    "block": BlockParser(),
+    "replacement": ReplacementParser(),
+    "split-diff": SplitDiffParser(),
+}
+
+
+class Session:
+    def __init__(
+        self,
+        stream: SessionStream,
+        config: ConfigManager,
+        code_context: CodeContext,
+        parser: Parser,
+    ):
+        self.stream = stream
+        self.config = config
+        self.code_context = code_context
+        self.parser = parser
+
+        self.id = uuid4()
+        self.parser = BlockParser()
+        self.code_file_manager = CodeFileManager(self.config)
+        self.cost_tracker = CostTracker()
+        setup_api_key()
+
+        self._main_task: asyncio.Task[None] | None = None
+        self._stop_task: asyncio.Task[None] | None = None
+
+    @classmethod
+    async def create(
+        cls,
+        paths: List[Path] = [],
+        exclude_paths: List[Path] = [],
+        no_code_map: bool = False,
+        diff: Optional[str] = None,
+        pr_diff: Optional[str] = None,
+    ):
+        stream = SessionStream()
+        await stream.start()
+        set_session_stream(stream)
+
+        git_root = get_shared_git_root_for_paths([Path(path) for path in paths])
+        # TODO: Config should be created in the client (i.e., to get vscode settings) and passed to session
+        config = await ConfigManager.create(git_root)
```

Comment:
```suggestion
        git_root = get_shared_git_root_for_paths([Path(path) for path in paths])
        # TODO: Config should be created in the client (i.e., to get vscode settings) and passed to session
        # TODO: calculate git_root in Config creation?
        config = await ConfigManager.create(git_root)
```

Reply:
I don't think it makes sense for git root to be in the config in the first place; ideally we move it out sometime soon after this PR (when we change how we pass around all of the other singletons)